.travis.yml

```diff
@@ -1,5 +1,29 @@
 language: c
+sudo: required
+env:
+  - TARGET=local
+  - TARGET=docker OS=ubuntu-16.04
+  - TARGET=docker OS=fedora-25
+  - TARGET=docker OS=opensuse-leap
+services:
+  - docker
 before_install:
-  - sudo apt-get update
-  - sudo apt-get install libsqlite3-dev
-script: ./autogen.sh && make && make check && make distcheck
+  - | # before_install: local
+    if [ "${TARGET}" = "local"  ]; then
+      sudo apt-get update && sudo apt-get install libsqlite3-dev
+    fi
+  - | # before_install: docker
+    if [ "${TARGET}" = "docker" ]; then
+      sudo docker pull "cangjians/build-essential:${OS}"
+      sudo docker tag "cangjians/build-essential:${OS}" "cangjians/build-essential:latest"
```

Comment:
I'm not sure why you're tagging the image here.

Is it so that in the VM, the `cangjians/build-essential:${OS}` image gets known as `cangjians/build-essential:latest`, which allows you to use a single Dockerfile for libcangjie with `FROM cangjians/build-essential:latest`?

So in the task which has `OS=ubuntu-lts` then the `cangjians/build-essential:latest` used by the libcangjie build is in fact `cangjians/build-essential:ubuntu-lts`, but in the task that has `OS=fedora-rawhide` then the same `cangjians/build-essential:latest` name points in fact to the `cangjians/build-essential:fedora-rawhide` image?

Reply:
Yup.

The descendent image will need to define exact <image:tag> target (i.e. the `FROM` statement). As you may see in the docker file for libcangjie and etc, we're bound to use `cangjians/buind-essential:latest`.

What I did was manually tagging our target OS image (`cangjians/build-essential:${OS}`) as `cangjians/buind-essential:latest`. Hence we trick the travis environment to use whatever image we want it to use.

Reply:
That's a neat trick!

However, I'm not sure that's going to work well in our case.

One thing I would like is that we publish libcangjie images the same way we publish the build-essential ones.

That is, when running the tests for ibus-cangjie, I **do not** want to first have to build the libcangjie image (from one of the build-essetional images), then build the pycangjie image (from the just-built libcangjie image), then finally build and run the tests for ibus-cangjie.

I really want the ibus-cangjie tests to just pull the pycangjie image right away.

I can see two ways to do that:

1. keeping your trick with a single Dockerfile, we tag the build image as `cangjians/libcangjie:${OS}`, and [when the CI finishes on master, we push the just built image to the registry](https://docs.travis-ci.com/user/docker/#Pushing-a-Docker-Image-to-a-Registry);
2. we use one dockerfile for each `${OS}`;

The first solution means that we don't rebuild/push the libcangjie images when the build-essential one changes, it only gets rebuilt/pushed when pushing to master.

Which leads me towards the second option, which just generally sucks. We'd have multiple dockerfiles, all with the exact same contents, except for the `FROM` line. 😞

Reply:
I've tried that approach. But on the way I realized that if we're going that path, we'll have to have multiple docker files in all our repositories to fulfil the dependencies along the way. (i.e. `libcangjie/Dockerfile.fedora-latest`, `libcangjie/Dockerfile.ubuntu-latest`, `libcangjie/Dockerfile.???` ... `ibus-cangjie/Dockerfile.fedora-latest`, `ibus-cangjie/Dockerfile.ubuntu-latest`, `ibus-cangjie/Dockerfile.???` ...

Whereas in the current approach, all the multiple-OS monstrosity only happen in the build-essential repository.

Reply:
In long run, if we want to support more and more OS, the current approach is much easier to maintain.

Reply:
I agree having a single Dockerfile is better, and I don't want to maintain tons of them.